repo-updater: Use rate limiter when syncing changesets

ryanslade opened the pull request:
I decided to do the rate limiting in the syncer itself, not in the source. The reason for this is that if we add a rate limiter in the source we need to remember to implement it every time we add a new source and also won't know which operation actually need to be rate limited.

For now, we specifically just need to rate limit syncs that happen as part of background syncing.

The syncer doesn't know the "cost" of a sync though, so I added a new `LoadChangesetCost` method to `ChangesetSource`. Although it expands the interface, I think it clearly describes the minimum that a Source needs to implement in order to support changeset syncing. 

I also moved some of the sync methods out into functions so that we don't need to have a bunch of short lived syncers that are only used once.

unknwon commented on cmd/repo-updater/repos/bitbucketserver.go:
```diff
@@ -187,6 +187,12 @@ func (s BitbucketServerSource) LoadChangesets(ctx context.Context, cs ...*Change
+// LoadChangesetCost returns the API cost of loading a changeset
+func (s BitbucketServerSource) LoadChangesetCost() int {
+	// We make 4 api calls
+	return 4
```
I somewhat feel it would be very trick to keep this number up-to-date... put them in a central place feels more nature and manageable. WDYT?

mrnugget replied:
Yeah, this seems really error-prone, especially since GitHub's `cost` is based on the GraphQL query, which is a long distance from this method.

unknwon commented on cmd/repo-updater/repos/bitbucketserver.go:
```diff
@@ -187,6 +187,12 @@ func (s BitbucketServerSource) LoadChangesets(ctx context.Context, cs ...*Change
 	return nil
 }
 
+// LoadChangesetCost returns the API cost of loading a changeset
```
```suggestion
// LoadChangesetCost returns the API cost of loading a changeset from Bitbucket Server.
```

unknwon commented on enterprise/internal/campaigns/syncer.go:
```diff
@@ -406,48 +415,44 @@ func (s *ChangesetSyncer) computeSchedule(ctx context.Context) ([]scheduledSync,
 // SyncChangesets refreshes the metadata of the given changesets and
 // updates them in the database.
-func (s *ChangesetSyncer) SyncChangesets(ctx context.Context, cs ...*campaigns.Changeset) (err error) {
+func SyncChangesets(ctx context.Context, repoStore RepoStore, syncStore SyncStore, cf *httpcli.Factory, cs ...*campaigns.Changeset) (err error) {
+	return syncChangesets(ctx, repoStore, syncStore, cf, nil, cs...)
+}
+
+func syncChangesets(ctx context.Context, repoStore RepoStore, syncStore SyncStore, cf *httpcli.Factory, rl *rate.Limiter, cs ...*campaigns.Changeset) (err error) {
```
It's probably worthing mentioning when to use which `[Ss]yncChangesets` in docstrings. Just a bit unclear to me.

tsenart commented:
> I decided to do the rate limiting in the syncer itself, not in the source. The reason for this is that if we add a rate limiter in the source we need to remember to implement it every time we add a new source and also won't know which operation actually need to be rate limited.

A `Source` maps 1:1 with an external service type, so it doesn't seem like we'd be implementing that many sources that this would be a concern. Having rate limiting be a `Source` concern makes a lot of sense to me and it would prevent the expansion of the interface with these `Cost` methods.

mrnugget commented on cmd/repo-updater/repos/sources.go:
```diff
@@ -93,6 +93,9 @@ type ChangesetSource interface {
 	// them. If a Changeset could not be found on the source, it's included in
 	// the returned slice.
 	LoadChangesets(context.Context, ...*Changeset) error
+	// LoadChangesetCost return estimated cost of loading one changeset. It
```
In which unit do you define `cost`? It seems like the GitHubSource uses GraphQL "rate limit points" and Bitbucket Server uses requests?

mrnugget commented on cmd/repo-updater/repos/github.go:
```diff
@@ -236,6 +236,13 @@ func (s GithubSource) LoadChangesets(ctx context.Context, cs ...*Changeset) erro
+func (s GithubSource) LoadChangesetCost() int {
+	// The cost can be requested from the GitHub API here:
+	// https://developer.github.com/v4/guides/resource-limitations/#returning-a-calls-rate-limit-status
+	return 3
```
So the cost here is the "point cost for the current call", meaning: a single request can have `cost > 1`.

How does this work with the rate limit being specified in `requestsPerHour`?

ryanslade replied:
I've refactored to do the rate limiting inside the source. 

I used "requestsPerHour" as that's the term most people are used to. In the case of GitHub when using the graphql api it'll use points instead.

mrnugget commented:
> Having rate limiting be a `Source` concern makes a lot of sense to me and it would prevent the expansion of the interface with these `Cost` methods.

I agree with this. If this were implemented on the source, the rate-limiting logic would be defined exactly where the requests are happening (if I understand it correctly).

And isn't this what we'd have to do anyway in order to cover `CreateChangeset` and the other methods anyway?

ryanslade commented:
> > Having rate limiting be a `Source` concern makes a lot of sense to me and it would prevent the expansion of the interface with these `Cost` methods.
> 
> I agree with this. If this were implemented on the source, the rate-limiting logic would be defined exactly where the requests are happening (if I understand it correctly).
> 
> And isn't this what we'd have to do anyway in order to cover `CreateChangeset` and the other methods anyway?

I've moved rate limiting into the source, but only implemented for the changeset operations. We technically only need it for loading changesets at the moment since that is all the syncer needs but I added it for all the `ChangesetSource` methods.

unknwon commented on cmd/repo-updater/repos/bitbucketserver.go:
```diff
@@ -101,6 +110,10 @@ func (s BitbucketServerSource) CreateChangeset(ctx context.Context, c *Changeset
 	pr.FromRef.Repository.Project.Key = repo.Project.Key
 	pr.FromRef.ID = git.EnsureRefPrefix(c.HeadRef)
 
+	if err := s.rateLimiter.Wait(ctx); err != nil {
```
qq: In what circumstances that this `Wait` method returns an error?

unknwon commented on cmd/repo-updater/repos/bitbucketserver.go:
```diff
@@ -188,6 +207,10 @@ func (s BitbucketServerSource) LoadChangesets(ctx context.Context, cs ...*Change
 }
 
 func (s BitbucketServerSource) loadPullRequestData(ctx context.Context, pr *bitbucketserver.PullRequest) error {
+	if err := s.rateLimiter.WaitN(ctx, 3); err != nil {
```
I think it is worthing document all required information verbosely to get this number 3. Therefore anyone later comes to the codebase can understand and/or fix it when 3 is no longer the "correct" value. WDYT?

unknwon replied:
ping @ryanslade in case you missed the this.

ryanslade replied:
Ah, sorry, I thought you meant that you were happy with it. I'll add a comment

unknwon commented on cmd/repo-updater/repos/github.go:
```diff
@@ -222,6 +237,12 @@ func (s GithubSource) LoadChangesets(ctx context.Context, cs ...*Changeset) erro
 		}
 	}
 
+	// Each LoadPullRequest call uses 3 tokens. This was calculated by manually calling the query
+	// and asking for the cost as described here:
+	// https://developer.github.com/v4/guides/resource-limitations/#returning-a-calls-rate-limit-status
```
This kind of info is sufficient to address my concern in the comment above.

unknwon commented on cmd/repo-updater/repos/sources.go:
```diff
@@ -52,13 +53,31 @@ func NewSourcer(cf *httpcli.Factory, decs ...func(Source) Source) Sourcer {
+	if err != nil {
+		return nil, err
+	}
+	css, ok := source.(ChangesetSource)
```
![](https://media1.giphy.com/media/m2YYMHxFFHwM8/giphy.gif?cid=5a38a5a2e74319d251a6083d8dfdaa2f70369f1f9c4a10a2&rid=giphy.gif)

unknwon commented on cmd/repo-updater/repos/sources.go:
```diff
@@ -52,13 +53,31 @@ func NewSourcer(cf *httpcli.Factory, decs ...func(Source) Source) Sourcer {
+	}
+	css, ok := source.(ChangesetSource)
+	if !ok {
+		return nil, fmt.Errorf("ChangesetSource cannot be created from external service %q", svc.Kind)
```
This seems new, would this cause the same confusion as in https://github.com/sourcegraph/sourcegraph/issues/9779?

ryanslade replied:
This time it's a valid error as we are trying to explicitly create a changeset source. This is only called from within the changeset syncer.
